src/H5C.c

```diff
@@ -7251,11 +7734,45 @@ H5C__load_entry(H5F_t *f,
 #ifdef H5_HAVE_PARALLEL
                         if (!coll_access || 0 == mpi_rank) {
 #endif /* H5_HAVE_PARALLEL */
-                            /* If the thing's image needs to be bigger for a speculatively
-                             * loaded thing, go get the on-disk image again (the extra portion).
+
+                            /* the original version of this code re-read
+                             * the entire buffer.  At some point, someone
+                             * reworked this code to avoid re-reading the
+                             * initial portion of the buffer.
+                             *
+                             * In addition to being of questionable utility,
+                             * this optimization changed the invariant that
+                             * that metadata is read and written atomically.
+                             * While this didn't cause immediate problems,
+                             * the page buffer in VFD SWMR depends on this
+                             * invariant in its management of multi-page
+                             * metadata entries.
+                             *
+                             * To repair this issue, I have reverted to
+                             * the original algorithm for managing the
+                             * speculative load case.  Note that I have
+                             * done so crudely -- before merge, we should
+                             * remove the infrastructure that supports the
+                             * optimization.
+                             *
+                             * We should also verify my impression that the
+                             * that the optimization is of no measurable
+                             * value.  If it is, we will put it back, but
+                             * disable it in the VFD SWMR case.
+                             *
+                             * While this issue was detected in the global
+                             * heap case, note that the super block, the
+                             * local heap, and the fractal heap also use
+                             * speculative loads.
+                             *
+                             *                          JRM -- 3/24/20
                              */
-                            if (H5F_block_read(f, type->mem_type, addr + len, actual_len - len, image + len) <
-                                0) {
+
+                            H5C__SET_PB_READ_HINTS(f->shared->cache, type, FALSE);
```

Comment:
There are enough of these little blocks of H5C__SET_PB_READ_HINTS -> H5F_block_read -> H5C__RESET_PB_READ_HINTS that I would consider creating a small routine to encapsulate them.   It's less likely that future maintenance will miss wrapping new calls to H5F_block_read or accidentally unwrap an existing one.